### PR TITLE
feat(header): addresing accessibility concerns

### PR DESCRIPTION
**Pull request by mnajdova**
# Header

This PR is introduced in order to address the accessibility concerns about the component. In the scope of this PR, the subheaderAs property was added for the Header Component, in order for the user to be able to change the element type of the Subheader component.

## Before
The structure of the rendered HTML before this PR was the following:
```html
<h1 class="ui-header a b">Account Settings
  <div class="ui-header__subheader c a b">Manage your account settings and set email preferences</div>
</h1>
```
## After
After introducing this changes the structure looks like this:
```html
<h1 class="ui-header a b">Account Settings</h1>
<p class="ui-header__subheader c a b">Manage your account settings and set email preferences</p>
```

## Concerns:
Although we have come up with the desired HTML structure proposed by @kuzhelov, I am concern about the way that we have the implementation now, because the styles root element is now applied only to the content of the header, as we got read of the wrapper element. If we replace the root with some other object from the styles, for example header, we will not have the default class by default, so I wasn't sure if this is what we wanted. Apart from this, using the child API will returned the previous implementation of the default shorthand usage.
I'd be happy to hear some proposals and comments for addressing this issues.
PS: I will apply the changes in the CHANGELOG.md once we agree on the previous issues

### CODE REVIEW

File: src/components/Header/Header.tsx [hunk 1]
```diff
@@ -30,11 +30,27 @@ class Header extends UIComponent<any, any> {
...
+    'className',
+    'content',
+    'subheader',
+    'subheaderAs',
+    'textAlign',
+  ]
+
+  static defaultProps = {
+    as: 'h1',
+    subheaderAs: 'h2',
```
Comment (levithomason) [inline src/components/Header/Header.tsx]: Use only the `subheader` prop.  All component parts (like subheader) should have a shorthand prop of the same name.  Shorthand allows passing values, but also configuring props where the user can define the `as` for the component part:

```jsx
<Header
  as='h1'
  content='Hi'
  subheader={{ as: 'h2', content: 'hello' }}
/>
```

Comment (mnajdova) [reply src/components/Header/Header.tsx]: Thanks for this info. I will make sure to include one example showing the possibility for customizing the subcomponents as well.

**Comment by mnajdova** [comment]
# Update 
After some discussions, I concluded the following. Introducing the `h2` as in the proposed solution would semantically mean starting new section in the content regarding the `h1` header. Because of this, I will revert the changes to the previous implementation with just replacing the `div` with default `small` property for the subheader. Maybe we should consider in the future changing the name of the subcomponent from subheader to `desription` or `additionalDescription`. Any suggested names for the subcomponent are welcome :) @kuzhelov @jurokapsiar @levithomason feel free to leave any comments/suggestions.

**Comment by jurokapsiar** [comment]
Agree with Marija's comment - h1 and h2 should form the structure of the page. In your case, subheader is more a description of the header, so the name should change. If the user wants to add ``<h1>`` and ``<h2>`` he would need to use two instances of the header component.

**Comment by mnajdova** [comment]
One last concern I have about this... By replacing the `div` with `small` tag, the subheader goes next to the header, instead of below it. So if semantically, the div was not problem and we want to stick with the initial look (to have the subheader below the header) then maybe we should have stayed with the div in the first place. (Will revert for now the subheader default behaviour to be div). Any thoughts?

## The implementation with div and using the as property for the subheader as span
image

**Comment by kuzhelov** [comment]
actually, even if we would like to use block-positioned element for subheading (instead of inlined `small`), I would recommend to use `p` instead of `div`, as it is more semantically correct in our context. Also may suggest to take a look on the following article that I come up to, it seems to describe exactly the case we are dealing with: http://html5doctor.com/howto-subheadings/

Here are key takeaways from it:
- h1 ... h6 tags must not be used for subheading (sorry, my bad)
- either of the next two approaches should be considered
```html
<h1>Header</h1>
<p>Subheader</p>
```
```html
<h1>
  Header <small>subheader</small>
</h1>
```
--------------
While choosing from these two, I'd put my vote in favor of the first one, because
- it has a markup that is more straightforward to the client
   - and, thus, it is more straightforward to manage by 'as' property provided (somehow) to subheader component - the effect of this will be predictable
- DOM elements are segregated

Also, if we would like to ensure specific positioning for the subheader (like that it will be rendered under the header), we might specify `display: block` property explicitly in its styles. With that there won't be any surprises to the client if she will decide to change default subheader's `p` to, say, `span` or `small`.

Please, let me know what do you think.

**Comment by jurokapsiar** [comment]
Thinking about this again - the subheader as defined is definitely not a h2. However, if the subheader is meant to be a description, the correct HTML (including container) should be:
```
<div role="application" aria-labelledby="calendar" aria-describedby="info">
    <h1 id="calendar">Calendar</h1>
    <p id="info">
        This calendar shows the game schedule for the Boston Red Sox.
    </p>
    <div role="grid">
        ...
    </div>
</div>
```
(as described in https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-describedby_attribute).

The correct solution should be somewhere in between the two approaches - to have 2 sibling elements, but not ``<h1>`` and ``<h2>`` but rather ``<h1>`` and ``<p>``. Note that ``<p>`` has restrictions around what elements can be inside of it, so the user might need to use a different element instead of it.
I think this brings a different concern - if we render two elements in a component, how do we make sure that these are referenceable from outside. Will this be possible by using following example?
```
<div aria-labelledby='header' aria-describedby='header-description<
  <Header id='header'>
    <Header.Description as='div' role='header' id='header-description'>
      Manage your account settings and set email preferences
    </Header.Description>
    Account Settings
  </Header>
  <p>Settings UI</p>
</div>
```

**Comment by kuzhelov** [comment]
So, it seems that we are in agreement here. As part of this PR we will introduce the following HTML structure:

```html
<h1>Header</h1>
<p>Subheader</p>
```

Speaking of whether the example @jurokapsiar is possible to achieve now - yes, it is. Please, @jurokapsiar, take another look on this

**Review by kuzhelov — APPROVED** [review]
@levithomason, please, take a look on that as well

**Comment by levithomason** [comment]
# Markup

If we can avoid returning siblings from the render function, that would be ideal.  Returning siblings will give us difficulties in layout as there is no containing element to position or style.

More importantly, it breaks the concept of "component parts" being "subcomponents" within a "component".  Breaking this pattern means there is no `root` element to style.  It has implications for conformance tests, shorthand props, and our public API.  Example:

```jsx
// valid :)
<Header subheader='' />

// invalid? :(
<Header>
  <Header.Subheader />
</Header>

// correct? :/
<Header>
  <Header.Header />
  <Header.Subheader />
</Header>
```

# Description vs Subheader

I really like this proposal and have shared the same thoughts for sometime.  Making this change would make the library more consistent an reduce our API surface area.

There are many components that have a main description area followed by a less prominent and more detailed description.  Currently, SUIR calls these title or header and subheader or description depending on where and how they appear.  I'd like to normalize the concepts if we can while maintaining intuitiveness.

Here are some examples we need to think about:

image

That is the Card.  The top bold/gray part could be the header/description.  However, we'd need to then name the text inside the Card something else.  Perhaps the `content` or the `main`?

***

image

That is a List with items that have a `header` and a `description`.  This could stay exactly how it is, both in naming and location of the parts.

One strange thing here is when a list item doesn't have a header and it just has some text.  In that case, we're currently using the `content` prop/part for the item.  This makes sense from a design and intuition standpoint, but the `content` and the `description` then are really doing the exact same thing.  This might pose an issue.

***

image

In all of these cases, there is the shorter part with the main info and a more detailed and less prominent description part.  I think all of these relationships could be represented as a `header`, `description`, and `content`.

**Comment by kuzhelov** [comment]
agreed to rename `subheader` to `description` and merge

**Comment by mnajdova** [comment]
I renamed the subheader to description and updated the CHANGELOG.md I think this should be ready to be merged now.

**Pull request closed without merging**